### ISSUE TITLE
Raptor: prevent nodes from filling-up when balancing

### ISSUE THREAD

**Issue by ggreg** [prestodb/presto#7062]
tl;dr Raptor does not know about non-assigned buckets and assigning buckets to a node that was non-available and then became available may lead to filling up its disk.

Let's consider a node becomes unavailable long enough to have all its buckets assigned to other nodes that are available (either when `SELECT` queries need to access its buckets or after the `BucketBalancer` has run). A good example is when there is a long network partition, for example when the rack switch is defective. The node keeps all the bucket files on its disk.

Then after some time, the node comes back and is available again. This node will have no buckets assigned to it in the Raptor metadata. When the `BucketBalancer` will run it will consider this node as empty and then will try to assign it many buckets to more evenly distribute the buckets across all nodes. The node will copy buckets from buckets either when queries need these buckets or when the shard recovery manager will find that the buckets assigned to this node are missing.

If the shard cleaner does not run in this interval of time (depending on `raptor.local-cleaner-interval`), or it waits before deleting the file of buckets no longer assigned to the node (depending on `raptor.local-clean-time`), the total size of all bucket files stored on disk is very likely to exceed the space available on disk (likely to be roughly twice the size that a normal node stores).

`raptor.local-clean-time` is used to prevent a race condition where the shard cleaner finds a shard that is being written and then not in the metadata yet. The shard cleaner would delete the file, i.e. eventually call `unlink(2)`, and the file will not exist once all the open files descriptors are closed (here when the writer finished writing the file and closed it). The problem can also be seen at the broader scope of the transaction that inserts multiple shards in the metadata.

The network partition case it interesting because the Presto process will not restart, so we cannot rely only on scanning non-assigned shards at startup/initialization time. However scanning non-assigned shards on startup covered most of the cases. It however relies on a external health check mechanism that stops Presto on the worker node if it is no longer available.

**Comment by ggreg** [comment]
I'll add the scanning of non-assigned shards in the `@PostConstruct` method to mark non-assigned shards and make the cleaner delete them.

**Comment by ggreg** [comment]
Discussed offline with @electrum and @erichwang. We can leverage the `created_shards` and `shards` (i.e. *assigned*) metadata tables to only clean deleted files. The local shard cleaner would then check `shards` first, then `created_shards`, and finally `shards` again.

**Comment by ggreg** [comment]
@jessesleeping did you already address some of this points?

**Comment by jessesleeping** [comment]
The concern expressed in this issue still exist. We changed the logic in shard cleaner a while ago to trigger a full clean up when the node is started. But in case of a network partition events, since the node is not restarted, the stated issue may still occur.
